File: 0029-divide-two-integers/0029-divide-two-integers.js
const divide = function(dividend, divisor) {
  if (dividend === 0) return 0;
  const sign = dividend < 0 && divisor > 0 || dividend > 0 && divisor < 0 ? -1 : 1;
  if (dividend < 0) dividend = -dividend;
  if (divisor < 0) divisor = -divisor;
  if (dividend < divisor) return 0;

  dividend = (dividend + '').split('').reverse();
  
  let res = '', part = dividend.pop(), count, left = '', zeroes = '';
  while (true) {
    if (+part >= divisor) {
      count = 0;
      while (+part >= divisor) {
        part -= divisor;
        ++count;
      }
      res += count;
      left = part = '' + part;
      zeroes = '';
    } else if (dividend.length) {
      if (!+left && +part) res += 0;
      part += dividend.pop();
      zeroes += '0';
    } else {
      res += zeroes;
      break;
    }
  }

  // fix for the test error
  if (sign > 0 && res > 2147483647) res = 2147483647;

  return sign < 0 ? -res : +res;
};

File: 0029-divide-two-integers/0029-divide-two-integers.go
import (
  "math"
  "strconv"
  "strings"
)

func divide(dividend int, divisor int) int {
  sign := dividend & math.MinInt32 ^ divisor & math.MinInt32
  if dividend < 0 { dividend = -dividend }
  if divisor < 0 { divisor = -divisor }
  if dividend == 0 || dividend < divisor { return 0 }

  dividendArr := []rune(strconv.Itoa(dividend))
  dividend, dividendArr = int(dividendArr[0] - '0'), dividendArr[1:]
  remainder, zeroes, ansArr := 0, 0, []rune{}

  for {
    if (dividend >= divisor) {
      num := 0
      for dividend >= divisor {
        dividend -= divisor
        num++
      }
      ansArr = append(ansArr, rune(num) + '0')
      remainder = dividend
      zeroes = 0
    } else if len(dividendArr) > 0 {
      if remainder == 0 && dividend > 0 {
        ansArr = append(ansArr, '0')
      }
      dividend, _ = strconv.Atoi(strconv.Itoa(dividend) + string(dividendArr[0]))
      dividendArr = dividendArr[1:]
      zeroes++
    } else {
      ansArr = append(ansArr, []rune(strings.Repeat("0", zeroes))...)
      break
    }
  }

  ans, _ := strconv.Atoi(string(ansArr))
  if sign != 0 { ans = -ans }
  if ans < math.MinInt32 { ans = math.MinInt32 }
  if ans > math.MaxInt32 { ans = math.MaxInt32 }
  return ans
}